### PR TITLE
feat: implement CSRF

### LINKED ISSUE #12: CSRF Protection
Shield is more or less completed, except the CSRF protection. This issue attempts to explain the requirements for CSRF Protection.

> One can look into the existing implementation of CSRF [here](https://github.com/adonisjs/shield/blob/v1.0.8/src/Shield/index.js).

- [ ] Standard implementation of CSRF that uses sessions for the secret and tokens for each form submission.
- [ ] View helper to dump CSRF token within the forms.
- [ ] View helper to set CSRF token as a meta tag. Helpful for Ajax requests.
- [ ] Setting the `XSRF-TOKEN` cookie.
- [ ] Accept the CSRF token using `X-CSRF-TOKEN` and `X-XSRF-TOKEN` headers

I think, we can also look into the referrer checks. Django does it, so I believe there will be some reason behind it

### PR DESCRIPTION
**Pull request by bahdcoder**
Install the csrf package, create csrf middleware function, handle disabled csrf cases, add tests for
csrf generation on every request'

Fix #12

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce?

_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] I have read the [CONTRIBUTING](https://github.com/adonisjs/adonis-shield/blob/master/CONTRIBUTING.md) doc
- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works.
- [X] I have added necessary documentation (if appropriate)

## Further comments

Some points to notice about this PR:

- The @adonisjs/auth package is not ready yet, and we need to save the csrf token into the session. For now I placed a basic object so tests can pass, but what do you think is a more permanent solution?
- I am also extracting the csrf token from `request.input('_csrf')` even though it was not specified in the specs. I'm hoping this is required functionality ?
- How are Http Exceptions handled in v5 ? I'm doing it the v4 way. Please have a look.
- I tried to find a better typescript typing for request methods, but couldn't. Maybe someone knows one ? `methods?: ReadonlyArray<string>,`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,6 +41,8 @@
     "typescript": "^3.8.3"
   },
   "dependencies": {
+    "@adonisjs/generic-exceptions": "^3.0.1",
```
Comment (thetutlage) [inline package.json]: We should be using `@poppinss/utils` and use the `Exception` class exported by it. The `@adonisjs/generic-exceptions` is depreciated

File: src/csrf.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import { RequestContract } from '@ioc:Adonis/Core/Request'
+import { HttpException } from '@adonisjs/generic-exceptions'
+import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'
+
+import { noop } from './noop'
+
+const Csrf = new Tokens()
+
+// TODO: Switch this to fetch secret from session once questions have been answered.
+const session: { secret?: string } = {}
```
Comment (thetutlage) [inline src/csrf.ts]: You don't need the authentication package for this. `@adonisjs/session` is used to deal with sessions

File: src/csrf.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  const method = request.method().toLowerCase()
+
+  if (! options.methods || options.methods.length === 0) {
+    return true
+  }
+
+  return options.methods.filter(definedMethod => definedMethod.toLowerCase() === method).length > 0
+}
+
+export function csrf (options: CsrfOptions, applicationKey: string) {
```
Comment (thetutlage) [inline src/csrf.ts]: Can we convert all methods above this function to a proper class? It will help structure the code a bit

Comment (bahdcoder) [reply src/csrf.ts]: Alright, on it!

**Comment by bahdcoder** [comment]
@thetutlage This is ready for review now. 

I have a question about the `path-to-regexp` package. I installed it because the `request.match` method was removed from the core. Do you think this is okay ?

Also, What do you think about changing the `filterUris` option to `except` ? I think using `except` can automatically make sense to a developer even without reading the documentation.

**Comment by thetutlage** [comment]
There is no need for `path-to-regexp` package. You can access the current route pattern from `ctx.route.pattern` as done inside the bodyparser https://github.com/adonisjs/bodyparser/blob/develop/src/BodyParser/index.ts#L137

> Also, What do you think about changing the filterUris option to except ? I think using except can automatically make sense to a developer even without reading the documentation?

Sure, `except` is more descriptive. How about `exceptRoutes`, I want to clear the intent, that they are defining routes and not URLs

File: test/csrf.spec.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    const csrfMiddleware = await getCsrfMiddlewareInstance({ enabled: true }, APPLICATION_SECRET_KEY)
+
+    const TEST_CSRF_SECRET = await csrfMiddleware.getCsrfSecret()
+    const TEST_CSRF_TOKEN = csrfMiddleware.generateCsrfToken(TEST_CSRF_SECRET)
+
+    const ctx = await getCtxFromIncomingMessage({
+      'x-csrf-token': TEST_CSRF_TOKEN,
+    })
+
+    ctx.session.put('csrf-secret', csrfMiddleware.session.get('csrf-secret'))
```
Comment (thetutlage) [inline test/csrf.spec.ts]: What this line is intended to do?

File: src/csrf.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+    if (this.requestMethodShouldEnforceCsrf(request) && this.requestUrlShouldEnforceCsrf(request)) {
+      const csrfToken = this.getCsrfTokenFromRequest(request)
+
+      if (! csrfToken || ! Csrf.verify(csrfSecret, csrfToken)) {
+        throw new Exception('Invalid CSRF Token', 403, 'E_BAD_CSRF_TOKEN')
+      }
+    }
+
+    request.csrfToken = this.generateCsrfToken(csrfSecret)
```
Comment (thetutlage) [inline src/csrf.ts]: We need to share this value with the view and also create `XSRF-TOKEN` cookie

File: test/csrf.spec.ts (hunk 4 above)
Comment (bahdcoder) [reply test/csrf.spec.ts]: To test that the csrf token is correctly verified, first I created a token, and while creating a token, a new csrf secret is generated. Next, I share that secret with the second context I generated. That way they both have the same secret and the verification can go through.

File: package.json [hunk 6]
```diff
@@ -20,16 +20,25 @@
...
+    "@adonisjs/session": "^2.3.3",
+    "@adonisjs/view": "^1.0.10",
+    "@poppinss/dev-utils": "^1.0.4",
+    "@types/csrf": "^1.3.2",
     "@types/node": "^13.7.7",
     "commitizen": "^4.0.3",
     "cz-conventional-changelog": "^3.1.0",
     "del-cli": "^3.0.0",
     "doctoc": "^1.4.0",
+    "edge.js": "^2.1.0",
```
Comment (thetutlage) [inline package.json]: The dev dependency has to be on `@adonisjs/view`

File: src/csrf.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  public generateCsrfToken (csrfSecret): string {
+    return Csrf.create(csrfSecret)
+  }
+
+  /**
+   * Set the xsrf cookie on
+   * response
+   */
+  private setXsrfCookie (ctx: HttpContextContract): void {
+    ctx.response.cookie('x-xsrf-token', ctx.request.csrfToken)
```
Comment (thetutlage) [inline src/csrf.ts]: The cookie name has to be `xsrf-token` without the leading `x`. Also, the `X-` is always a convention for header keys.

File: src/csrf.ts [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+   * This would make a csrfToken variable available to
+   * the edge view templates. This would also create
+   * a helpful method called csrfField to be used
+   * on the frontend to generate a hidden
+   * field called _csrf
+   */
+  private shareCsrfViewLocals (ctx: HttpContextContract): void {
+    ctx.view.share({
+      csrfToken: ctx.request.csrfToken,
+      csrfMeta: (compilerContext) => compilerContext.safe(`<meta name='csrf-token' content='${ctx.request.csrfToken}'>`),
```
Comment (thetutlage) [inline src/csrf.ts]: Love that you figured out the `safe` method 👌

File: test/csrf.spec.ts (hunk 4 above)
Comment (thetutlage) [reply test/csrf.spec.ts]: Okay, makes sense :)

File: src/csrf.ts [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+  /**
+   * This would make a csrfToken variable available to
+   * the edge view templates. This would also create
+   * a helpful method called csrfField to be used
+   * on the frontend to generate a hidden
+   * field called _csrf
+   */
+  private shareCsrfViewLocals (ctx: HttpContextContract): void {
+    ctx.view.share({
```
Comment (thetutlage) [inline src/csrf.ts]: Just add a conditional here to check that `ctx.view` is defined. In some rare edge cases, it is possible that someone is used CSRF but without the views, or maybe they swapped the template engine to something else

**Comment by thetutlage** [comment]
Thanks 😄

**Pull request merged**